### PR TITLE
ENH: `broadcast_shapes`

### PR DESCRIPTION
**Pull request by crusaderky**
`lazy_apply` benefits from this.

### CODE REVIEW

File: src/array_api_extra/_lib/_funcs.py [hunk 1]
```diff
@@ -71,6 +72,67 @@ def atleast_nd(x: Array, /, *, ndim: int, xp: ModuleType | None = None) -> Array
     return x
 
 
+def broadcast_shapes(*shapes: tuple[float | None, ...]) -> tuple[int | None, ...]:
```
Comment (lucascolley) [inline src/array_api_extra/_lib/_funcs.py:75]:
```suggestion
def broadcast_shapes(*shapes: tuple[int | None, ...]) -> tuple[int | None, ...]:
```

File: tests/test_funcs.py [hunk 2]
```diff
@@ -113,6 +115,62 @@ def test_xp(self, xp: ModuleType):
...
+        with pytest.raises(ValueError, match=match):
+            _ = broadcast_shapes(*args)
+
+    @pytest.mark.parametrize(
+        "args",
+        [
+            ((None,), (None,)),
+            ((math.nan,), (None,)),
+            ((1, None, 2, 4), (2, 3, None, 1), (2, None, None, 4)),
+            ((1, math.nan, 2), (4, 2, 3, math.nan), (4, 2, None, None)),
```
Comment (lucascolley) [inline tests/test_funcs.py:165]: Not that there is a library for which you would hit this in practice, but we could add a test where the input has _both_ `None` and `math.nan` dimensions? Feel free to ignore if that seems pointless.

**Review by lucascolley — APPROVED** [review]
thanks @crusaderky, LGTM!

File: src/array_api_extra/_lib/_funcs.py (hunk 1 above)
Comment (crusaderky) [reply src/array_api_extra/_lib/_funcs.py:75]: Not a typo. It is to allow for Dask's NaNs. Alas `Literal[math.nan]` is not a valid type.

File: tests/test_funcs.py (hunk 2 above)
Comment (crusaderky) [reply tests/test_funcs.py:165]: Good point. One may use the value returned by a previous call to `broadcast_shapes` with a shape of a dask array. Added it.

File: src/array_api_extra/_lib/_funcs.py (hunk 1 above)
Comment (lucascolley) [reply src/array_api_extra/_lib/_funcs.py:75]: Do we want `float | int | None` then?

Comment (crusaderky) [reply src/array_api_extra/_lib/_funcs.py:75]: in typing (but not at runtime), float is a superclass of int https://peps.python.org/pep-0484/#the-numeric-tower

Comment (lucascolley) [reply src/array_api_extra/_lib/_funcs.py:75]:
```suggestion
# `float` in signature to accept `math.nan` for Dask.
# `int`s are still accepted as `float` is a superclass of `int` in typing
def broadcast_shapes(*shapes: tuple[float | None, ...]) -> tuple[int | None, ...]:
```
